src/algebra/lie_algebra.lean

```diff
@@ -175,28 +175,120 @@ end prio
 
 namespace lie_algebra
 
-/--
-A morphism of Lie algebras is a linear map respecting the bracket operations.
--/
-structure morphism (R : Type u) (L : Type v) (L' : Type v)
+set_option old_structure_cmd true
+/-- A morphism of Lie algebras is a linear map respecting the bracket operations. -/
+structure morphism (R : Type u) (L : Type v) (L' : Type w)
   [comm_ring R] [lie_ring L] [lie_algebra R L] [lie_ring L'] [lie_algebra R L']
   extends linear_map R L L' :=
 (map_lie : ∀ {x y : L}, to_fun ⁅x, y⁆ = ⁅to_fun x, to_fun y⁆)
 
 infixr ` →ₗ⁅⁆ `:25 := morphism _
 notation L ` →ₗ⁅`:25 R:25 `⁆ `:0 L':0 := morphism R L L'
 
-instance (R : Type u) (L : Type v) (L' : Type v)
-    [comm_ring R] [lie_ring L] [lie_algebra R L] [lie_ring L'] [lie_algebra R L'] :
-  has_coe (L →ₗ⁅R⁆ L') (L →ₗ[R] L') := ⟨morphism.to_linear_map⟩
+section morphism_properties
+
+variables {R : Type u} {L₁ : Type v} {L₂ : Type w} {L₃ : Type w₁}
+variables [comm_ring R] [lie_ring L₁] [lie_ring L₂] [lie_ring L₃]
+variables [lie_algebra R L₁] [lie_algebra R L₂] [lie_algebra R L₃]
+
+instance : has_coe (L₁ →ₗ⁅R⁆ L₂) (L₁ →ₗ[R] L₂) := ⟨morphism.to_linear_map⟩
+
+lemma map_lie (f : L₁ →ₗ⁅R⁆ L₂) (x y : L₁) : f ⁅x, y⁆ = ⁅f x, f y⁆ := morphism.map_lie f
```

Comment:
This is also a simp-lemma, right?

Reply:
In fact no, because the LHS is not in simp normal form, and so this could never fire as a simp lemma.

I only know this because I had essentially the same thought and looked at the history to discover how things got the way they are and found https://github.com/leanprover-community/mathlib/pull/2066/files#diff-3e1340faac56a2d8f9fbc96b584c147a

If you tag it with `@[simp]` and run the linter, it says:
```
/- SOME SIMP LEMMAS ARE REDUNDANT.
That is, their left-hand side is not in simp-normal form.
These lemmas are hence never used by the simplifier.

This linter gives you a list of other simp lemmas, look at them!

Here are some guidelines to get you started:

  1. 'the left-hand side reduces to XYZ':
     you should probably use XYZ as the left-hand side.

  2. 'simp can prove this':
     This typically means that lemma is a duplicate, or is shadowed by another lemma:

     2a. Always put more general lemmas after specific ones:

      @[simp] lemma zero_add_zero : 0 + 0 = 0 := rfl
      @[simp] lemma add_zero : x + 0 = x := rfl

      And not the other way around!  The simplifier always picks the last matching lemma.

     2b. You can also use @[priority] instead of moving simp-lemmas around in the file.

      Tip: the default priority is 1000.
      Use `@[priority 1100]` instead of moving a lemma down,
      and `@[priority 900]` instead of moving a lemma up.

     2c. Conditional simp lemmas are tried last, if they are shadowed
         just remove the simp attribute.

     2d. If two lemmas are duplicates, the linter will complain about the first one.
         Try to fix the second one instead!
         (You can find it among the other simp lemmas the linter prints out!)
: -/
#print lie_algebra.map_lie /- Left-hand side simplifies from
  ⇑f ⁅x,y⁆
to
  ⇑↑f ⁅x,y⁆
using 
  [coe_fn_coe_base]
Try to change the left-hand side to the simplified term!
 -/
```

Reply:
Ugh! Do we know why `⇑f` is simplifying to `⇑↑f` in the first place? That doesn't look like much of a simplification to me. :-(

Reply:
This is on purpose.  If you have a coercion from `ℕ` to `ℤ` and a second coercion from `ℤ` to `ℚ`, then for `x : ℕ` you can still write `(x : ℚ)`.  Lean will "transparently" chain these coercions and produce the term `↑x` (yes, one single arrow!).  (This is probably going to be just as bad in Lean 4.)

However you don't want to write lemmas for all possible combinations of coercions.  Therefore transitive coercions are simplified to multiple coercions.

Let me quote the "function coercion" library note:

Many structures such as bundled morphisms coerce to functions so that you can
transparently apply them to arguments. For example, if `e : α ≃ β` and `a : α`
then you can write `e a` and this is elaborated as `⇑e a`. This type of
coercion is implemented using the `has_coe_to_fun`type class. There is one
important consideration:

If a type coerces to another type which in turn coerces to a function,
then it **must** implement `has_coe_to_fun` directly:
```lean
structure sparkling_equiv (α β) extends α ≃ β

-- if we add a `has_coe` instance,
instance {α β} : has_coe (sparkling_equiv α β) (α ≃ β) :=
⟨sparkling_equiv.to_equiv⟩

-- then a `has_coe_to_fun` instance **must** be added as well:
instance {α β} : has_coe_to_fun (sparkling_equiv α β) :=
⟨λ _, α → β, λ f, f.to_equiv.to_fun⟩
```

(Rationale: if we do not declare the direct coercion, then `⇑e a` is not in
simp-normal form. The lemma `coe_fn_coe_base` will unfold it to `⇑↑e a`. This 
often causes loops in the simplifier.)

Reply:
Thanks, @gebner, I definitely learnt something there!

This rule presumably causes a problem when we have diamonds of coercions, but I guess we don't face that often.